llvm/test/CodeGen/AArch64/hadd-combine.ll

```diff
@@ -859,6 +859,21 @@ define <4 x i32> @urhadd_v4i32(<4 x i32> %x) {
   ret <4 x i32> %r
 }
 
+define <4 x i32> @fixedwidth(<4 x i32> %a0, <4 x i32> %a1)  {
+; CHECK-LABEL: fixedwidth:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    and v2.16b, v0.16b, v1.16b
+; CHECK-NEXT:    eor v0.16b, v0.16b, v1.16b
+; CHECK-NEXT:    usra v2.4s, v0.4s, #1
+; CHECK-NEXT:    mov v0.16b, v2.16b
+; CHECK-NEXT:    ret
```

Comment:
This failed to match the uhadd instruction? I think because of the isVector test

Reply:
I think we should implement the pattern logic in TargetLowering.cpp only. Because for vector implementation, we need data like : 
```
const APInt &DemandedBits,
const APInt &DemandedElts,
```
Which are defined in the latter but would need to be added in the former.

Reply:
Why do you think you need the demanded masks for vectors?

Reply:
My understanding of implementing vector type is limited to `TargetLowering.cpp` for now in which we have used the demanded masks for vectors while in `DAGCombiner.cpp` we have simply returned SDValue(). But I suppose we can implement vectors using this? : 

```cpp
  EVT NVT = EVT::getIntegerVT(*DAG.getContext(), llvm::bit_ceil(VT.getScalarSizeInBits()));
  if (VT.isVector())
    NVT = EVT::getVectorVT(*DAG.getContext(), NVT, VT.getVectorElementCount());
```

Reply:
You shouldn't need any EVT conversion - the whole point of the fixedwidth patterns is that its the same type, with no extension/truncation.

Reply:
Well yes, but isn't integer VT different than vector VT? Please pardon my silly questions. From my understanding, FixedWidth means not extending the bits as we would have done in ADD to prevent overflow.

Reply:
Resolved. Tests are passing now.